Create a vector indexing roachtest

mw5h opened the pull request:
#### workload/vecann: add shared utility functions for vector workloads

Add DeriveDistanceMetric and CalculateRecall functions to the vecann
package to be shared between vecbench and the upcoming vector index
roachtest. These functions extract the distance metric from dataset
naming conventions and compute recall for search result validation.

Refactor vecbench to use these shared implementations, removing
duplicate code.

Informs: #154590
Release note: None

#### roachtest/vecindex: add a vector index test stub

Add minimal test infrastructure for vector index roachtests:
- Create vecindex.go with vecIndexOptions struct and empty
  registerVectorIndex function
- Register the test in registry.go
- Add helper functions for test naming, key generation, and distance
  operator mapping.
- Stubs out the phases of the test.

Add test configurations and registration for 5 vector index test
variants:
- vecindex/dbpedia-100k/nodes=3 (standard, no prefix)
- vecindex/dbpedia-100k/nodes=3/prefix=3 (with prefix columns)
- vecindex/dbpedia-1m/nodes=6 (large-scale)
- vecindex/random-s/nodes=1 (local development)
- vecindex/random-s/nodes=1/prefix=2 (local with prefix)

Informs: #154590
Release note: None

#### roachtest/vecindex: implement a test of backfill and merge

Add a test phase that loads a dataset using a pool of workers and, at a
test-specified percentage of table population, kicks off a create vector
index for the data. This allows us to test both backfill (pre-create)
and merge (post-create starting). Times are reported for both but are
not used as a pass criteria (yet).

Informs: #154590
Release note: None

#### roachtest/vecindex: test recall of vector ann data

This addition to the vecindex roachtest tests the recall of nearest
neighbors from the test data provided with each data set. Each test has
a configurable set of beam sizes to test and a minimum recall
correctneess that is acceptable for each beam size. Tests that load
multiple prefixes test each prefix.

Informs: #154590
Release note: None

#### roachtest/vecindex: add a concurrent reader/writer subtest

This subtest spins up a configurable number of readers and writers to
drive vector search load to the database.

Each writer inserts rows from the first train data file in single
row batches until it has inserted all of the rows in that file, at
which point it switches into delete mode and starts deleting rows
in 10 row batches. When all rows have been deleted, the writer once
again becomes an inserter and the process repeats.

Each reader randomly selects a beam size from the sizes configured for
the test and then runs searches for random vectors in the test data for
the dataset. The reader ignores rows inserted by the writer threads to
avoid too heavily skewing results. To do this, it searches for more
vectors than called for and then filters the output to remove vectors
written by the insert workers. When the read worker exits, it validates
its recall rate against the expected rate for the number of searches it
performed.

For multi-prefix tests, this subtest only reads and writes to the first
prefix to ensure the maximum amount of contention.

Fixes: #154590
Release note: None

mw5h committed:
workload/vecann: add shared utility functions for vector workloads

Add DeriveDistanceMetric and CalculateRecall functions to the vecann
package to be shared between vecbench and the upcoming vector index
roachtest. These functions extract the distance metric from dataset
naming conventions and compute recall for search result validation.

Refactor vecbench to use these shared implementations, removing
duplicate code.

Informs: #154590
Release note: None

mw5h committed:
roachtest/vecindex: add a vector index test stub

Add minimal test infrastructure for vector index roachtests:
- Create vecindex.go with vecIndexOptions struct and empty
  registerVectorIndex function
- Register the test in registry.go
- Add helper functions for test naming, key generation, and distance
  operator mapping.
- Stubs out the phases of the test.

Add test configurations and registration for 5 vector index test
variants:
- vecindex/dbpedia-100k/nodes=3 (standard, no prefix)
- vecindex/dbpedia-100k/nodes=3/prefix=3 (with prefix columns)
- vecindex/dbpedia-1m/nodes=6 (large-scale)
- vecindex/random-s/nodes=1 (local development)
- vecindex/random-s/nodes=1/prefix=2 (local with prefix)

Informs: #154590
Release note: None

mw5h committed:
roachtest/vecindex: implement a test of backfill and merge

Add a test phase that loads a dataset using a pool of workers and, at a
test-specified percentage of table population, kicks off a create vector
index for the data. This allows us to test both backfill (pre-create)
and merge (post-create starting). Times are reported for both but are
not used as a pass criteria (yet).

Informs: #154590
Release note: None

mw5h committed:
roachtest/vecindex: test recall of vector ann data

This addition to the vecindex roachtest tests the recall of nearest
neighbors from the test data provided with each data set. Each test has
a configurable set of beam sizes to test and a minimum recall
correctneess that is acceptable for each beam size. Tests that load
multiple prefixes test each prefix.

Informs: #154590
Release note: None

mw5h committed:
roachtest/vecindex: add a concurrent reader/writer subtest

This subtest spins up a configurable number of readers and writers to
drive vector search load to the database.

Each writer inserts rows from the first train data file in single
row batches until it has inserted all of the rows in that file, at
which point it switches into delete mode and starts deleting rows
in 10 row batches. When all rows have been deleted, the writer once
again becomes an inserter and the process repeats.

Each reader randomly selects a beam size from the sizes configured for
the test and then runs searches for random vectors in the test data for
the dataset. The reader ignores rows inserted by the writer threads to
avoid too heavily skewing results. To do this, it searches for more
vectors than called for and then filters the output to remove vectors
written by the insert workers. When the read worker exits, it validates
its recall rate against the expected rate for the number of searches it
performed.

For multi-prefix tests, this subtest only reads and writes to the first
prefix to ensure the maximum amount of contention.

Fixes: #154590
Release note: None

mw5h commented:
Friendly ping @nameisbhaskar @shailendra-patel. This tests 25.4 GA functionality, so it'd be good to get it running.

DarrylWong reviewed (commented):
I'm mostly concerned that this test is doing too much on the test runner. One (unideal) solution is to provision an extra node and scp/run the code there - that's what some of the cdc roachtests do.

I guess a potentially naive question is: since we already have a vecann workload, why can't this be implemented in the same way, i.e. as a CLI command?

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> sans questions about separating the test runner from the DB machine.

> @DrewKimball reviewed 3 of 3 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155998#-:-Ocr6Ji4CXDguJa3Nljm:b-fx5kpb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mw5h, @nameisbhaskar, and @shailendra-patel)



<!-- Sent from Reviewable.io -->

mw5h committed:
roachtest/vecindex: create a nightly test for vector indexing

The new vector index roachtest doest too much work on the test runner.
To avoid flakiness, we're going to run this test on a separate test
runner until a more roachtest-friendly version of the test can be built.
These scripts implement that test.

Informs: #154590
Release note: None

mw5h commented:
@DarrylWong I think this is RFAL.

DarrylWong reviewed (approved):
LGTM!
